core/src/input/keymap.rs

```diff
@@ -105,131 +106,174 @@ where
     /// let keymap = Keymap::<Action>::new();
     /// ```
     pub fn new() -> Self {
-        Self { bindings: HashMap::new() }
+        Self { actions: HashMap::new() }
     }
 
-    /// Inserts or overwrites a keybinding of the keymap.
+    /// Inserts an entry into the keymap.
     ///
-    /// # Examples
+    /// This method is for internal use only.
+    /// To insert an entry into the keymap at runtime use the [`KeymapEvent::InsertAction`] event.
+    fn insert(&mut self, action: T, chord: KeyChord) -> &mut Self {
+        if let Some(actions) = self.actions.get_mut(&chord) {
+            if !actions.contains(&action) {
+                actions.push(action);
+            }
+        } else {
+            self.actions.insert(chord, vec![action]);
+        }
+        self
+    }
+
+    /// Removes an entry of the keymap.
     ///
-    /// ```
-    /// # use vizia_core::*;
-    /// #
-    /// # #[derive(PartialEq, Eq, Hash, Copy, Clone)]
-    /// # enum Action {
-    /// #     One,
-    /// # }
-    /// #
-    /// let keymap = Keymap::new()
-    ///     .insert(Action::One, KeyBinding::new(Modifiers::CTRL, Code::KeyA));
-    /// ```
-    pub fn insert(mut self, key: T, binding: KeyBinding) -> Self {
-        self.bindings.insert(key, binding);
+    /// This method is for internal use only.
+    /// To remove an entry of the keymap at runtime use the [`KeymapEvent::RemoveAction`] event.
+    fn remove(&mut self, action: &T, chord: &KeyChord) -> &mut Self {
+        if let Some(actions) = self.actions.get_mut(chord) {
+            if let Some(index) = actions.iter().position(|x| x == action) {
+                if actions.len() == 1 {
+                    self.actions.remove(chord);
+                } else {
+                    actions.swap_remove(index);
+                }
+            }
+        }
         self
     }
 
-    /// Removes a keybinding of the keymap if it exists.
+    /// Returns an iterator over every pressed action or `None` if there are no actions for that key chord.
     ///
     /// # Examples
     ///
     /// ```
     /// # use vizia_core::*;
     /// #
-    /// # #[derive(PartialEq, Eq, Hash, Copy, Clone)]
+    /// # #[derive(Debug, PartialEq, Copy, Clone)]
     /// # enum Action {
     /// #     One,
     /// # }
     /// #
-    /// let keymap = Keymap::new().remove(&Action::One);
+    /// # let cx = &Context::new();
+    /// # let keymap = Keymap::<Action>::new();
+    /// #
+    /// if let Some(actions) = keymap.pressed_actions(cx, Code::KeyA) {
+    ///     for action in actions {
+    ///         println!("The action {:?} is being pressed!", action);
+    ///     }
+    /// };
     /// ```
-    pub fn remove(mut self, key: &T) -> Self {
-        self.bindings.remove(key);
-        self
+    pub fn pressed_actions(&self, cx: &Context, code: Code) -> Option<impl Iterator<Item = &T>> {
```

Comment:
Why is this an Option<impl Iterator> and not just an iterator which is sometimes empty?

Reply:
That's because I'm returning an iterator over `&T` which in turn doesn't allow me to return something like `Vec::new().iter()`, because that value would be dropped at the end of the scope. Is there a way to get around this that you know of?

Reply:
Can you do something like `actions.get(...).iter().flat_map(|actions| actions.iter())`? Option has an iter method which iterates over the single item if it's Some, and nothing if it's None.

Reply:
I couldn't make your suggestion work, because it would return a `FlatMap` but I found another solution is just to use [].iter() in the else case ^^ Now it just returns an `impl Iterator<Item = &T>` without an `Option` around it.